flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -477,6 +478,50 @@ KeyGroupedInternalPriorityQueue<T> create(
         }
     }
 
+    @Override
+    public <N, S extends State, V> S upgradeKeyedState(
+            TypeSerializer<N> namespaceSerializer, StateDescriptor<S, V> stateDescriptor)
+            throws Exception {
+        StateFactory stateFactory = getStateFactory(stateDescriptor);
+        Tuple2<ColumnFamilyHandle, RegisteredKeyValueStateBackendMetaInfo<N, V>> registerResult =
+                tryRegisterKvStateInformation(stateDescriptor, namespaceSerializer, noTransform());
+        Preconditions.checkState(kvStateInformation.containsKey(stateDescriptor.getName()));
+        kvStateInformation.computeIfPresent(
+                stateDescriptor.getName(),
+                (stateName, kvStateInfo) ->
+                        new RocksDbKvStateInfo(
+                                kvStateInfo.columnFamilyHandle,
+                                new RegisteredKeyValueStateBackendMetaInfo<>(
+                                        kvStateInfo.metaInfo.snapshot())));
+        return stateFactory.createState(
+                stateDescriptor, registerResult, RocksDBKeyedStateBackend.this);
```

Comment:
As discussed offline, can the creation of state object be avoided?

ditto: `ChangelogRestoreTarget.createKeyedState`, `HeapKeyedStateBackend.upgradeKeyedState`

Reply:
I found it's a bit difficult to avoid creating a new state object while upgrading (I just limit the number of creations).
Because in current machanism of applier, a state object is needed to insert when applier.
Do you have any suggestions ?

Reply:
I see the following options:
1. Create new objects (as it is done currently in the PR)
2. Facilitate upgrade by state objects (e.g. `RocksDBValueState`) - i.e. add methods to update the serializer (and make the field mutable)
3. Use getter (`RegisteredKeyValueStateBackendMetaInfo.getStateSerializer`) in state objects (e.g. ` RocksDBValueState`) instead of injecting it
4. Wrap serializer to delegate all accesses to it and allow updating the actual serializer "externally" (similar to what was initially described in the ticket)

2nd option looks better to me because it is explicit, more simple than others and doesn't incur any overhead.
WDYT?

Reply:
Thanks a lot for the suggestion! I also think 2nd option seems better.
Besides this problem, It could also solve the next problem you mentioned below.
Because what I concerned in the next problem is how to update the serializer when one serializer COMPATIBLE_AS_IS with the other.
I will try to modify it using the 2nd option.

Reply:
Would also be interesting to hear @Myasuka opinion.

We were discussing updating value serializers in state objects here, like setting `valueSerializer` in `AbstractRocksDBState`. This would help in state migration when using changelog state backend.
It is different from the current case, when state objects are created only **after** recovery; with changelog enabled, state objects are created **during** recovery.

Reply:
What do you think about the object creation in `upgradeKeyedState` and in `createInternalState` ?
What we may all aggre is that we should create the object only once during recovery ?
If we make it better, the object could just be created once during recovery and user access, we may need to store created InternalKvState in `RocksDBKeyedStateBackend` and `HashMapKeyedStateBackend`. But It may modify more codes in these classes.
WDYT?

Reply:
1. What is the problem with creating new objects? Is it because of making too many objects or resource leaks?
This is the most clean to me if neither is the case.

2. Second option is fine as well. One question: if we make it mutable, is it possible from user code that updates the serializer during runtime (job running) and causes unexpected updates?

Reply:
1. I have the following concerns regarding re-creating state objects:
a) Old objects might leak some resources
b) New objects might not inherit all the "meta' state (such as currentNamespace, changelogger.metadataFlag), or even state updates
c) Re-creating state objects might not be enough; in particular, for HeapStateBackend we might also want to update the unerlying `CopyOnWriteStateMap` 
Probably, those are not immediate issue per se, but at the very least they make updating serializers error-prone for future changes.

2. Good point, I think unexpected updates and migrations can be prevented by formalizing the lifecycle of state objects (and checking whether the current phase allows migration or not)

Maybe we don't need to make the fields non-final, but make the lifecycle explicit, e.g. by adding `RocksDbValueState.withNewSerializer(...)` if it works better.

WDYT?

Reply:
If we wanna to create the object only once, not only we need to define a method like RocksDbValueState.withNewSerializer(...)  as roman said,
but also we need to store the created InternalKvState in RocksDBKeyedStateBackend / HashMapKeyedStateBackend, and modify the existing logic of createInternalState (to check whether there are created InternalKvState).
I think it could work to avoid something unexpected while creating the object more than once as roman said.
But it does need to modify some existing logic, so maybe @Myasuka could share some ideas about it ?

Reply:
I don't think we should add such APIs in the public user-facing interfaces, can we add this `withNewSerializer(...)` interface in the `InternalKvState`, which should be internal only.

Reply:
> I don't think we should add such APIs in the public user-facing interfaces, can we add this `withNewSerializer(...)` interface in the `InternalKvState`, which should be internal only.

OK, I am with this. I think this is a better place to go.

Reply:
I agree.

Reply:
I think we are all on the same page now and it makes sense to document this design somehwere, along with other aspects of this feature - for future reference (as well as to make sure we're indeed on the same page).
WDYT?

Reply:
Sure, I will also try to document the design and share it in the ticket.